// near-protocol-rewards/dist/src/index.js
"use strict";
/**
 * NEAR Protocol Rewards SDK
 *
 * Track and reward developer contributions through GitHub metrics
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorCode = exports.BaseError = exports.RateLimiter = exports.ConsoleLogger = exports.GitHubCollector = exports.GitHubRewardsSDK = void 0;
var sdk_1 = require("./sdk");
Object.defineProperty(exports, "GitHubRewardsSDK", { enumerable: true, get: function () { return sdk_1.GitHubRewardsSDK; } });
var github_1 = require("./collectors/github");
Object.defineProperty(exports, "GitHubCollector", { enumerable: true, get: function () { return github_1.GitHubCollector; } });
var logger_1 = require("./utils/logger");
Object.defineProperty(exports, "ConsoleLogger", { enumerable: true, get: function () { return logger_1.ConsoleLogger; } });
var rate_limiter_1 = require("./utils/rate-limiter");
Object.defineProperty(exports, "RateLimiter", { enumerable: true, get: function () { return rate_limiter_1.RateLimiter; } });
var errors_1 = require("./types/errors");
Object.defineProperty(exports, "BaseError", { enumerable: true, get: function () { return errors_1.BaseError; } });
Object.defineProperty(exports, "ErrorCode", { enumerable: true, get: function () { return errors_1.ErrorCode; } });
